mongodbatlas/mongodbatlas.go

```diff
@@ -233,29 +250,71 @@ func SetUserAgent(ua string) ClientOpt {
 // BaseURL of the Client. Relative URLS should always be specified without a preceding slash. If specified, the
 // value pointed to by body is JSON encoded and included in as the request body.
 func (c *Client) NewRequest(ctx context.Context, method, urlStr string, body interface{}) (*http.Request, error) {
-	rel, err := url.Parse(urlStr)
+	u, err := c.BaseURL.Parse(urlStr)
 	if err != nil {
 		return nil, err
 	}
 
-	u := c.BaseURL.ResolveReference(rel)
+	buf, err := c.ReadBody(body)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(method, u.String(), buf)
+	if err != nil {
+		return nil, err
+	}
 
-	buf := new(bytes.Buffer)
 	if body != nil {
-		err = json.NewEncoder(buf).Encode(body)
+		req.Header.Set("Content-Type", mediaType)
+	}
+	req.Header.Add("Accept", mediaType)
+	if c.UserAgent != "" {
+		req.Header.Set("User-Agent", c.UserAgent)
+	}
+	return req, nil
+}
+
+// ReadBody returns an ReadWriter object containing the body for the http request
+func (c *Client) ReadBody(body interface{}) (io.ReadWriter, error) {
```

Comment:
[nit] couple of things, this is not "reading a body as much as it is encoding it or serialising it, which would be a more correct action, this is the body of the client request, not from the server response
I guess you created the method to reuse it in both NewRequest and NewGZipRequest, like I mention before `NewGZipRequest` should not need this, so you could delete this method
If you decide to keep it please consider to make it private and not public along with the name change